src/main/java/slimeknights/tconstruct/common/data/CastingRecipeBuilder.java

```diff
@@ -0,0 +1,189 @@
+package slimeknights.tconstruct.common.data;
+
+import com.google.gson.JsonObject;
+import net.minecraft.advancements.Advancement;
+import net.minecraft.advancements.AdvancementRewards;
+import net.minecraft.advancements.ICriterionInstance;
+import net.minecraft.advancements.IRequirementsStrategy;
+import net.minecraft.advancements.criterion.RecipeUnlockedTrigger;
+import net.minecraft.data.IFinishedRecipe;
+import net.minecraft.item.Item;
+import net.minecraft.item.crafting.IRecipeSerializer;
+import net.minecraft.item.crafting.Ingredient;
+import net.minecraft.tags.Tag;
+import net.minecraft.util.IItemProvider;
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.registries.ForgeRegistries;
+import slimeknights.tconstruct.library.recipe.RecipeUtil;
+import slimeknights.tconstruct.smeltery.TinkerSmeltery;
+import slimeknights.tconstruct.smeltery.recipe.AbstractCastingRecipe;
+
+import javax.annotation.Nullable;
+import java.util.function.Consumer;
+
+public class CastingRecipeBuilder {
```

Comment:
Look into `AbstractRecipeBuilder`, it handles the advancement builder for you. Extend it like `class CastingRecipeBuilder extends AbstractRecipeBuilder<CastingRecipeBuilder>`, and in your build, call that one funny method that returns a resource location. Let me know if you have questions, or see something else that would be good in the abstract builder.